src: refactor coverage connection

joyeecheung opened the pull request:
#### src: refactor coverage connection

- Refactor the C++ class to be resuable for other types of profiles
- Move the try-catch block around coverage collection callback
  to be inside the callback to silence potential JSON or write
  errors.
- Use Function::Call instead of MakeCallback to call the coverage
  message callback since it does not actually need async hook
  handling. This way we no longer needs to disable the async
  hooks when writing the coverage results.
- Renames `lib/internal/coverage-gen/with_profiler.js` to
  `lib/internal/profiler.js` because it is now the only way
  to generate coverage.

#### test: show stderr on v8 coverage test failures

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

joyeecheung commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/21322/

addaleax commented on src/inspector_profiler.cc:
```diff
@@ -0,0 +1,216 @@
+  }
+
+  void StartProfiling() override {
+    Debug(this->env(),
```
```suggestion
    Debug(env(),
```

(here and below)

gengjiawen commented on src/inspector_profiler.cc:
```diff
@@ -0,0 +1,216 @@
+          "Sending Profiler.startPreciseCoverage\n");
+    Isolate* isolate = this->env()->isolate();
+    Local<String> enable = FIXED_ONE_BYTE_STRING(
+        isolate, "{\"id\": 1, \"method\": \"Profiler.enable\"}");
```
Use raw string here ?

gengjiawen commented on src/inspector_profiler.cc:
```diff
@@ -0,0 +1,216 @@
+};
+
+void StartCoverageCollection(Environment* env) {
+  V8CoverageConnection* connection = new V8CoverageConnection(env);
```
auto here, https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-auto.html#new-expressions ?

joyeecheung replied:
In this case our style guide takes precedence because it's a simple type: https://github.com/nodejs/node/blob/master/CPP_STYLE_GUIDE.md#using-auto

> `Being explicit about types is usually preferred over using auto.`

gengjiawen commented on src/inspector_profiler.cc:
```diff
@@ -0,0 +1,216 @@
+  virtual void EndProfiling() = 0;
+  virtual void SetConnection(Local<Object> connection) = 0;
+
+  friend class V8ProfilerSessionDelegate;
```
Is this declaration redundant ?

joyeecheung commented:
Rebased and addressed reviews. CI: https://ci.nodejs.org/job/node-test-pull-request/21434/

@gengjiawen Regarding raw strings: I have no idea why but it looks like neither we nor non-test V8 code use those, so I'd prefer leaving that out of this PR, in case there's any weird compiler support oddities. Feel free to open a separate PR to update those.

joyeecheung commented:
<img width="1081" alt="Screen Shot 2019-03-12 at 6 59 47 AM" src="https://user-images.githubusercontent.com/4299420/54163639-eec24d00-4459-11e9-8520-b81f6c5110de.png">

I don't think there is a job to run the coverage CI against a PR. I do get a successful run locally though. If this comes back breaking the coverage, we can just revert.

joyeecheung commented:
Landed in 963ee0b...eb0bf8d

BridgeAR commented:
This does not land cleanly on v11. It seems to rely on other commits that should be backported first. Please open a manual backport for it or change the labels accordingly.
